google/cloud/spanner/internal/partial_result_set_source.cc

```diff
@@ -14,186 +14,243 @@
+      // Copy the column names into a vector that will be shared with
       // every Row object returned from NextRow().
       columns_ = std::make_shared<std::vector<std::string>>();
+      columns_->reserve(metadata_->row_type().fields_size());
       for (auto const& field : metadata_->row_type().fields()) {
         columns_->push_back(field.name());
       }
     }
   }
-
   if (result_set->result.has_stats()) {
-    // If we got stats more than once, log it, but use the last one.
+    // If we get stats more than once, log it, but use the last one.
     if (stats_) {
-      GCP_LOG(WARNING) << "Unexpectedly received two sets of stats";
+      GCP_LOG(WARNING) << "PartialResultSetSource: Additional stats";
     }
     stats_ = std::move(*result_set->result.mutable_stats());
   }
 
   // If reader_->Read() resulted in a new PartialResultSetReader (i.e., it
-  // used the last resume_token to resume an interrupted stream), then we
-  // must clear the buffered partial-row data as it will be replayed.
+  // used the token to resume an interrupted stream), then we must discard
+  // any buffered data as it will be replayed.
   if (result_set->resumption) {
-    buffer_.clear();
-    chunk_ = {};
-  }
-
-  auto& new_values = *result_set->result.mutable_values();
-
-  // Merge values if necessary, as described in:
-  // https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.PartialResultSet
-  //
-  // As an example, if we receive the following 4 responses (assume the values
-  // are all `string_value`s of type `STRING`):
-  //
-  // ```
-  // { { values: ["A", "B", "C1"] }  chunked_value: true }
-  // { { values: ["C2", "D", "E1"] } chunked_value: true }
-  // { { values: ["E2"] },           chunked_value: true }
-  // { { values: ["E3", "F"] }       chunked_value: false }
-  // ```
-  //
-  // The final values yielded are: `A`, `B`, `C1C2`, `D`, `E1E2E3`, `F`.
-  //
-  // n.b. One value can span more than two responses (the `E1E2E3` case above);
-  // the code "just works" without needing to treat that as a special-case.
-  if (chunk_) {
-    if (new_values.empty()) {
+    if (!resume_token_) {
+      // The reader claims to have resumed the stream even though we said it
+      // should not. That leaves us in the untenable position of possibly
+      // having returned data that will be replayed, so fail the stream now.
       return Status(StatusCode::kInternal,
-                    "PartialResultSet contained no values "
-                    "to merge with prior chunked_value");
+                    "PartialResultSetSource reader resumed the stream"
+                    " despite our having asked it not to");
     }
-    auto& front = new_values[0];
-    auto merge_status = MergeChunk(*chunk_, std::move(front));
-    if (!merge_status.ok()) {
-      return merge_status;
+    values_back_incomplete_ = false;
+    values_.Clear();
+  }
+
```

Comment:
I'm hesitant to because `PartialResultSetSource` doesn't really know anything about chunking.  Take a look at `merge_chunk.h` and `merge_chunk.cc` and let me know if you think moving the example there would have a benefit.  Thanks.

Reply:
I am just letting you know that the example was easier for me to grok than the prose. I think it's fine to remove the example, because there are [tests](https://github.com/googleapis/google-cloud-cpp/blob/6f512195c0a14a34ba7637f46d0e4e3f34a91179/google/cloud/spanner/internal/partial_result_set_source_test.cc#L525-L529) that do the same job as the example.

> `PartialResultSetSource` doesn't really know anything about chunking.

This is an odd claim when the logic depends on the `chunked_value`.... 🤷‍♂️

Reply:
> I am just letting you know that the example was easier for me to grok than the prose. I think it's fine to remove the example, because there are [tests](https://github.com/googleapis/google-cloud-cpp/blob/6f512195c0a14a34ba7637f46d0e4e3f34a91179/google/cloud/spanner/internal/partial_result_set_source_test.cc#L525-L529) that do the same job as the example.
> 
> > `PartialResultSetSource` doesn't really know anything about chunking.
> 
> This is an odd claim when the logic depends on the `chunked_value`.... man_shrugging

I guess what I'm saying is that, while it knows values can be split between responses, it doesn't know anything about how they are reconstituted.